numpy/typing/_dtype_like.py

```diff
@@ -1,10 +1,37 @@
-from typing import Any, Dict, List, Sequence, Tuple, Union
+import sys
+from typing import Any, List, Sequence, Tuple, Union, TYPE_CHECKING
 
 from numpy import dtype
 from ._shape import _ShapeLike
 
 _DtypeLikeNested = Any  # TODO: wait for support for recursive types
 
+if TYPE_CHECKING:
+    if sys.version_info >= (3, 8):
+        from typing import Protocol, TypedDict
+    else:
+        from typing_extensions import Protocol, TypedDict
+
+    # Mandatory keys
+    class _DtypeDictBase(TypedDict):
+        names: Sequence[str]
+        formats: Sequence[_DtypeLikeNested]
+
+    # Mandatory + optional keys
+    class _DtypeDict(_DtypeDictBase, total=False):
+        offsets: Sequence[int]
+        titles: Sequence[Any]  # Elements should be valid as dict keys
+        itemsize: int
+        aligned: bool
+
+    # A protocol for anything with the dtype attribute
+    class _SupportsDtype(Protocol):
+        dtype: _DtypeLikeNested
+
+else:  # runtime-only placeholders
+    _DtypeDict = 'numpy.typing._dtype_like._DtypeDict'
```

Comment:
We had previously decided against using runtime-only placeholders unless it was 100% necessary; i.e. `sys.version_info < (3, 8, )` and `typing_extensions` not installed. Why break from that here?

Reply:
> Why break from that here?

An oversight I'm afraid. Fixed in https://github.com/numpy/numpy/pull/16622/commits/fd796670d9b9cc14dc6a98a70cfe495e72e5691e.